tsdb/engine/tsm1/engine.go

```diff
@@ -223,50 +228,89 @@ func (e *DevEngine) Begin(writable bool) (tsdb.Tx, error) {
 
 func (e *DevEngine) WriteTo(w io.Writer) (n int64, err error) { panic("not implemented") }
 
-func (e *DevEngine) compact() {
-	for {
+func (e *DevEngine) writeSnapshot() error {
+	// Lock and grab the cache snapshot along with all the closed WAL
+	// filenames associated with the snapshot
+	closedFiles, snapshot, err := func() ([]string, *Cache, error) {
+		e.mu.Lock()
+		defer e.mu.Unlock()
+
 		if err := e.WAL.CloseSegment(); err != nil {
-			e.logger.Printf("error rolling current WAL segment: %v", err)
-			time.Sleep(time.Second)
-			continue
+			return nil, nil, err
 		}
 
-		tsmFiles, segments, err := e.CompactionPlan.Plan()
+		segments, err := e.WAL.ClosedSegments()
 		if err != nil {
-			e.logger.Printf("error calculating compaction plan: %v", err)
-			time.Sleep(time.Second)
-			continue
+			return nil, nil, err
 		}
 
-		if len(tsmFiles) == 0 && len(segments) == 0 {
+		snapshot := e.Cache.Snapshot()
+
+		return segments, snapshot, nil
+	}()
+
+	if err != nil {
+		return err
+	}
+
+	go func() {
```

Comment:
Does it need a read lock? It's not accessing any fields that get modified. Also, putting a read lock on it would eliminate the point of having it run in a goroutine since we wouldn't be able to kick off a snapshot write of a new Cache while this write is happening. (since there's a full Lock to start it)